File: services/extract/packet_splitter.py
"""Packet splitter — classifies a multi-document packet into typed sections.

An optional pipeline stage between chunking and routing that recognizes
when a single upload contains multiple logically distinct documents
(invoice + COI + policy stapled together, SEC filing + cover letter, etc.)
and splits them into contiguous chunk-range sections so each can be
extracted with the appropriate schema.

Design doc: docs/design/classify-split.md
"""

from __future__ import annotations

import json
import re
import time
from dataclasses import dataclass, field

from .document_map import Chunk
from .providers import ModelProvider

# Reserved type IDs — always valid even if the schema author didn't
# declare them in koji.yaml.
# - `other` is the catch-all the classifier can use when it doesn't
#   recognize a block as one of the declared types.
# - `document` is the fallback type written by the normalizer when the
#   classifier fails entirely (invalid JSON, every section rejected, etc.)
#   and the whole document collapses back to a single section.
_OTHER_TYPE_ID = "other"
_FALLBACK_TYPE_ID = "document"

_CHUNK_PREVIEW_CHARS = 400


@dataclass
class Section:
    """A contiguous chunk range classified as a single document type.

    The splitter produces one Section per distinct document it finds in
    a packet. Each section owns a contiguous range of chunk indices so
    downstream routing and extraction can operate on a slice of the
    original document_map without losing identity.
    """

    type: str
    title: str
    chunk_indices: list[int] = field(default_factory=list)
    confidence: float = 0.0

    @property
    def start_chunk(self) -> int:
        return self.chunk_indices[0] if self.chunk_indices else 0

    @property
    def end_chunk(self) -> int:
        return self.chunk_indices[-1] if self.chunk_indices else 0


def _build_classifier_prompt(chunks: list[Chunk], types: list[dict]) -> str:
    """Build the LLM prompt asking for direct section ranges."""
    type_lines = []
    for t in types:
        if not isinstance(t, dict):
            continue
        type_id = t.get("id")
        if not isinstance(type_id, str) or not type_id.strip():
            continue
        description = (t.get("description") or "").strip() or type_id
        type_lines.append(f"- {type_id}: {description}")
    type_lines.append(f"- {_OTHER_TYPE_ID}: Anything not matching the types above.")
    types_block = "\n".join(type_lines)

    chunk_block = "\n\n".join(f"[{c.index}] {c.title}\n{c.content[:_CHUNK_PREVIEW_CHARS]}" for c in chunks)

    return f"""You're given a document that may be a single item or a packet of several stapled-together documents. Identify each logical document in the packet and return its type and chunk range.

Types:
{types_block}

Rules:
- Each chunk belongs to exactly one section.
- Section ranges must be contiguous (start_chunk to end_chunk, inclusive).
- Sections must not overlap.
- Every chunk in the document must belong to some section — no gaps.
- If unsure about a block, use type "{_OTHER_TYPE_ID}" rather than inventing.

Return JSON in this exact shape:
{{
  "sections": [
    {{"type": "invoice", "start_chunk": 0, "end_chunk": 4, "confidence": 0.96}}
  ]
}}

Chunks to classify:

{chunk_block}
"""


def _parse_classifier_json(raw: str) -> dict | None:
    """Extract the JSON object from the classifier's raw response."""
    if not isinstance(raw, str) or not raw.strip():
        return None
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        match = re.search(r"\{[\s\S]*\}", raw)
        if match:
            try:
                return json.loads(match.group())
            except json.JSONDecodeError:
                return None
    return None


def _fallback_section(total_chunks: int) -> Section:
    """Single-section fallback covering every chunk.

    Used whenever the classifier fails or returns nothing usable — this
    is the "behave like classify was never enabled for this document"
    escape hatch.
    """
    return Section(
        type=_FALLBACK_TYPE_ID,
        title="Document (classifier fallback)",
        chunk_indices=list(range(total_chunks)),
        confidence=0.0,
    )


def _normalize_classifier_response(
    raw_response: dict | None,
    total_chunks: int,
    valid_type_ids: set[str],
) -> tuple[list[Section], int]:
    """Turn a classifier response into a clean validated list of Sections.

    Handles every failure mode the LLM response can produce:

    1. Invalid JSON / non-dict response: fall back to single section.
    2. Empty `sections` list: fall back.
    3. Section with missing / malformed fields: drop it.
    4. Out-of-range indices (start or end beyond the chunk count, or
       negative): drop the section.
    5. Inverted range (start > end): drop the section.
    6. Overlapping ranges: resolve with first-start-wins (the later
       section's overlap is trimmed off the front; if nothing remains,
       drop it entirely).
    7. Gaps between sections: fill with an `other` section so no chunk
       is orphaned.
    8. Unknown type IDs (type not in valid_type_ids or "other"):
       coerce to `other`.

    Returns (sections, corrections_count) where corrections_count tracks
    how many fixups the normalizer applied — useful observability for
    tuning prompts or spotting a classifier that's going off the rails.
    """
    corrections = 0

    if total_chunks <= 0:
        return ([], 0)

    if not isinstance(raw_response, dict):
        return ([_fallback_section(total_chunks)], 1)

    raw_sections = raw_response.get("sections")
    if not isinstance(raw_sections, list) or not raw_sections:
        return ([_fallback_section(total_chunks)], 1)

    # Step 1: validate each section in isolation.
    validated: list[dict] = []
    for entry in raw_sections:
        if not isinstance(entry, dict):
            corrections += 1
            continue

        stype = entry.get("type")
        start = entry.get("start_chunk")
        end = entry.get("end_chunk")
        conf_raw = entry.get("confidence", 0.0)

        if not isinstance(stype, str) or not stype.strip():
            corrections += 1
            continue
        if not isinstance(start, int) or not isinstance(end, int):
            corrections += 1
            continue
        if start < 0 or end < 0 or start >= total_chunks or end >= total_chunks:
            corrections += 1
            continue
        if start > end:
            corrections += 1
            continue

        if stype not in valid_type_ids and stype != _OTHER_TYPE_ID:
            corrections += 1
            stype = _OTHER_TYPE_ID

        try:
            conf = float(conf_raw)
        except (TypeError, ValueError):
            conf = 0.0
        conf = max(0.0, min(1.0, conf))

        validated.append(
            {
                "type": stype,
                "start": start,
                "end": end,
                "confidence": conf,
            }
        )

    if not validated:
        return ([_fallback_section(total_chunks)], corrections + 1)

    # Step 2: sort by start_chunk and resolve overlaps (first-start-wins).
    validated.sort(key=lambda s: (s["start"], s["end"]))
    cleaned: list[dict] = []
    last_end = -1
    for entry in validated:
        start = entry["start"]
        end = entry["end"]
        if start <= last_end:
            corrections += 1
            new_start = last_end + 1
            if new_start > end:
                # Entirely consumed by a previous section; drop it.
                continue
            entry = {**entry, "start": new_start}
        cleaned.append(entry)
        last_end = entry["end"]

    if not cleaned:
        return ([_fallback_section(total_chunks)], corrections + 1)

    # Step 3: fill gaps with "other" sections so every chunk is claimed.
    with_gaps: list[dict] = []
    cursor = 0
    for entry in cleaned:
        if entry["start"] > cursor:
            corrections += 1
            with_gaps.append(
                {
                    "type": _OTHER_TYPE_ID,
                    "start": cursor,
                    "end": entry["start"] - 1,
                    "confidence": 0.0,
                }
            )
        with_gaps.append(entry)
        cursor = entry["end"] + 1

    if cursor < total_chunks:
        corrections += 1
        with_gaps.append(
            {
                "type": _OTHER_TYPE_ID,
                "start": cursor,
                "end": total_chunks - 1,
                "confidence": 0.0,
            }
        )

    sections = [
        Section(
            type=entry["type"],
            title=f"Section {i} — {entry['type']}",
            chunk_indices=list(range(entry["start"], entry["end"] + 1)),
            confidence=entry["confidence"],
        )
        for i, entry in enumerate(with_gaps)
    ]
    return (sections, corrections)


async def classify_chunks_to_sections(
    chunks: list[Chunk],
    provider: ModelProvider,
    types: list[dict],
) -> tuple[list[Section], dict]:
    """Run the classifier and return (sections, metadata).

    metadata shape:
        - total_sections: int — final section count after normalization
        - elapsed_ms: int — wall time for the classifier call + normalizer
        - normalizer_corrections: int — number of fix-ups applied
        - error: str (optional) — present if the LLM call raised or JSON
          failed to parse, indicating the pipeline fell back to a single
          default section
    """
    start = time.time()

    if not chunks:
        return (
            [],
            {
                "total_sections": 0,
                "elapsed_ms": 0,
                "normalizer_corrections": 0,
            },
        )

    valid_type_ids = {
        t["id"] for t in types if isinstance(t, dict) and isinstance(t.get("id"), str) and t["id"].strip()
    }
    prompt = _build_classifier_prompt(chunks, types)

    raw: str | None = None
    error: str | None = None
    try:
        raw = await provider.generate(prompt, json_mode=True)
    except Exception as exc:  # noqa: BLE001
        error = f"{type(exc).__name__}: {exc}"
        print(f"[koji-classify] Classifier error: {error} — falling back to whole-document section")

    parsed = _parse_classifier_json(raw) if raw is not None else None
    sections, corrections = _normalize_classifier_response(parsed, len(chunks), valid_type_ids)

    metadata: dict = {
        "total_sections": len(sections),
        "elapsed_ms": int((time.time() - start) * 1000),
        "normalizer_corrections": corrections,
    }
    if error is not None:
        metadata["error"] = error
    return (sections, metadata)

File: api/src/extract/packet-splitter.ts
/**
 * Packet splitter — classifies a multi-document packet into typed sections.
 *
 * TypeScript port of services/extract/packet_splitter.py.
 *
 * An optional pipeline stage between chunking and routing that recognizes
 * when a single upload contains multiple logically distinct documents
 * and splits them into contiguous chunk-range sections so each can be
 * extracted with the appropriate schema.
 */

import type { Chunk } from "./document-map";
import type { ModelProvider } from "./providers";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Catch-all type the classifier uses for unrecognized blocks. */
export const OTHER_TYPE_ID = "other";

/** Fallback type when the classifier fails entirely or doc is too short. */
export const FALLBACK_TYPE_ID = "document";

const CHUNK_PREVIEW_CHARS = 400;

/** Documents at or below this chunk count skip the LLM classifier. */
export const DEFAULT_SHORT_DOC_CHUNKS = 2;

/**
 * Fraction of chunks a single non-other type must cover for the
 * post-classify coalesce step to fire.
 */
export const DEFAULT_COALESCE_OTHER_THRESHOLD = 0.5;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Section {
  type: string;
  startChunk: number;
  endChunk: number;
  chunks: Chunk[];
  confidence: number;
}

export interface ClassifyResult {
  sections: Section[];
  corrections: number;
  classifierSkipped: boolean;
}

interface ClassifyOptions {
  shortDocChunks?: number;
  coalesceOtherThreshold?: number;
}

// ---------------------------------------------------------------------------
// Prompt builder
// ---------------------------------------------------------------------------

function buildClassifierPrompt(chunks: Chunk[], types: Array<{ id?: string; description?: string }>): string {
  const typeLines: string[] = [];
  for (const t of types) {
    if (typeof t !== "object" || t === null) continue;
    const typeId = t.id;
    if (typeof typeId !== "string" || !typeId.trim()) continue;
    const description = (t.description ?? "").trim() || typeId;
    typeLines.push(`- ${typeId}: ${description}`);
  }
  typeLines.push(`- ${OTHER_TYPE_ID}: Anything not matching the types above.`);

  const chunkBlock = chunks
    .map((c) => `[${c.index}] ${c.title}\n${c.content.slice(0, CHUNK_PREVIEW_CHARS)}`)
    .join("\n\n");

  return `You're given a document that may be a single item or a packet of several stapled-together documents. Identify each logical document in the packet and return its type and chunk range.

Types:
${typeLines.join("\n")}

Rules:
- Each chunk belongs to exactly one section.
- Section ranges must be contiguous (start_chunk to end_chunk, inclusive).
- Sections must not overlap.
- Every chunk in the document must belong to some section — no gaps.
- If unsure about a block, use type "${OTHER_TYPE_ID}" rather than inventing.

Return JSON in this exact shape:
{
  "sections": [
    {"type": "invoice", "start_chunk": 0, "end_chunk": 4, "confidence": 0.96}
  ]
}

Chunks to classify:

${chunkBlock}
`;
}

// ---------------------------------------------------------------------------
// JSON parser (tolerant of markdown fences etc.)
// ---------------------------------------------------------------------------

function parseClassifierJson(raw: string): Record<string, unknown> | null {
  if (typeof raw !== "string" || !raw.trim()) return null;
  try {
    return JSON.parse(raw) as Record<string, unknown>;
  } catch {
    const match = raw.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        return JSON.parse(match[0]) as Record<string, unknown>;
      } catch {
        return null;
      }
    }
    return null;
  }
}

// ---------------------------------------------------------------------------
// Fallback section
// ---------------------------------------------------------------------------

function fallbackSection(chunks: Chunk[]): Section {
  return {
    type: FALLBACK_TYPE_ID,
    startChunk: 0,
    endChunk: Math.max(0, chunks.length - 1),
    chunks,
    confidence: 0,
  };
}

// ---------------------------------------------------------------------------
// Normalizer
// ---------------------------------------------------------------------------

/**
 * Turn a raw classifier response into validated Sections.
 *
 * Handles 8 failure modes — see packet_splitter.py docstring for details.
 */
export function normalizeClassifierResponse(
  rawResponse: Record<string, unknown> | null,
  totalChunks: number,
  validTypeIds: Set<string>,
  chunks?: Chunk[],
): [Section[], number] {
  let corrections = 0;

  if (totalChunks <= 0) return [[], 0];

  // Provide stub chunks if caller didn't pass them (for unit testing the normalizer directly)
  const resolvedChunks =
    chunks ??
    Array.from({ length: totalChunks }, (_, i) => ({
      index: i,
      title: `Chunk ${i}`,
      content: "",
      signals: { has_dates: false, has_dollar_amounts: false, has_tables: false, has_key_value_pairs: false },
      charOffset: 0,
      charLength: 0,
    }));

  if (rawResponse === null || typeof rawResponse !== "object" || Array.isArray(rawResponse)) {
    return [
      [fallbackSection(resolvedChunks)],
      1,
    ];
  }

  const rawSections = rawResponse.sections;
  if (!Array.isArray(rawSections) || rawSections.length === 0) {
    return [
      [fallbackSection(resolvedChunks)],
      1,
    ];
  }

  // Step 1: validate each section in isolation
  interface ValidatedEntry {
    type: string;
    start: number;
    end: number;
    confidence: number;
  }

  const validated: ValidatedEntry[] = [];

  for (const entry of rawSections) {
    if (typeof entry !== "object" || entry === null || Array.isArray(entry)) {
      corrections++;
      continue;
    }

    let stype = (entry as Record<string, unknown>).type;
    const start = (entry as Record<string, unknown>).start_chunk;
    const end = (entry as Record<string, unknown>).end_chunk;
    const confRaw = (entry as Record<string, unknown>).confidence ?? 0;

    if (typeof stype !== "string" || !stype.trim()) {
      corrections++;
      continue;
    }
    if (typeof start !== "number" || typeof end !== "number") {
      corrections++;
      continue;
    }
    if (start < 0 || end < 0 || start >= totalChunks || end >= totalChunks) {
      corrections++;
      continue;
    }
    if (start > end) {
      corrections++;
      continue;
    }

    if (stype !== OTHER_TYPE_ID && !validTypeIds.has(stype)) {
      corrections++;
      stype = OTHER_TYPE_ID;
    }

    let conf = typeof confRaw === "number" ? confRaw : parseFloat(String(confRaw));
    if (isNaN(conf)) conf = 0;
    conf = Math.max(0, Math.min(1, conf));

    validated.push({ type: stype, start, end, confidence: conf });
  }

  if (validated.length === 0) {
    return [
      [fallbackSection(resolvedChunks)],
      corrections + 1,
    ];
  }

  // Step 2: sort by start and resolve overlaps (first-start-wins)
  validated.sort((a, b) => a.start - b.start || a.end - b.end);

  const cleaned: ValidatedEntry[] = [];
  let lastEnd = -1;

  for (const entry of validated) {
    let { start } = entry;
    const { end } = entry;

    if (start <= lastEnd) {
      corrections++;
      const newStart = lastEnd + 1;
      if (newStart > end) continue; // entirely consumed
      start = newStart;
      cleaned.push({ ...entry, start });
    } else {
      cleaned.push(entry);
    }
    lastEnd = cleaned[cleaned.length - 1].end;
  }

  if (cleaned.length === 0) {
    return [
      [fallbackSection(resolvedChunks)],
      corrections + 1,
    ];
  }

  // Step 3: fill gaps with "other" sections
  const withGaps: ValidatedEntry[] = [];
  let cursor = 0;

  for (const entry of cleaned) {
    if (entry.start > cursor) {
      corrections++;
      withGaps.push({
        type: OTHER_TYPE_ID,
        start: cursor,
        end: entry.start - 1,
        confidence: 0,
      });
    }
    withGaps.push(entry);
    cursor = entry.end + 1;
  }

  if (cursor < totalChunks) {
    corrections++;
    withGaps.push({
      type: OTHER_TYPE_ID,
      start: cursor,
      end: totalChunks - 1,
      confidence: 0,
    });
  }

  const sections: Section[] = withGaps.map((entry) => ({
    type: entry.type,
    startChunk: entry.start,
    endChunk: entry.end,
    chunks: resolvedChunks.slice(entry.start, entry.end + 1),
    confidence: entry.confidence,
  }));

  return [sections, corrections];
}

// ---------------------------------------------------------------------------
// Coalescer
// ---------------------------------------------------------------------------

/**
 * Collapse other-sprinkled single-document output into one typed section.
 */
export function coalesceOtherSections(
  sections: Section[],
  totalChunks: number,
  threshold: number,
  allChunks?: Chunk[],
): [Section[], string | null] {
  if (threshold <= 0 || totalChunks <= 0 || sections.length === 0) {
    return [sections, null];
  }

  const hasOther = sections.some((s) => s.type === OTHER_TYPE_ID);
  if (!hasOther) return [sections, null];

  const chunksByType: Record<string, number> = {};
  const confByType: Record<string, number[]> = {};

  for (const s of sections) {
    if (s.type === OTHER_TYPE_ID || s.type === FALLBACK_TYPE_ID) continue;
    const count = s.endChunk - s.startChunk + 1;
    chunksByType[s.type] = (chunksByType[s.type] ?? 0) + count;
    confByType[s.type] = confByType[s.type] ?? [];
    confByType[s.type].push(s.confidence);
  }

  const typeKeys = Object.keys(chunksByType);
  if (typeKeys.length === 0) return [sections, null];

  let dominantType = typeKeys[0];
  for (const t of typeKeys) {
    if (chunksByType[t] > chunksByType[dominantType]) dominantType = t;
  }

  const dominantFraction = chunksByType[dominantType] / totalChunks;
  if (dominantFraction < threshold) return [sections, null];

  const confs = confByType[dominantType] ?? [0];
  const mergedConf = confs.reduce((a, b) => a + b, 0) / confs.length;

  const resolvedChunks =
    allChunks ??
    Array.from({ length: totalChunks }, (_, i) => ({
      index: i,
      title: `Chunk ${i}`,
      content: "",
      signals: { has_dates: false, has_dollar_amounts: false, has_tables: false, has_key_value_pairs: false },
      charOffset: 0,
      charLength: 0,
    }));

  const coalesced: Section = {
    type: dominantType,
    startChunk: 0,
    endChunk: totalChunks - 1,
    chunks: resolvedChunks,
    confidence: mergedConf,
  };

  return [[coalesced], dominantType];
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

export async function classifyChunksToSections(
  chunks: Chunk[],
  provider: ModelProvider,
  types: Array<{ id?: string; description?: string }>,
  options?: ClassifyOptions,
): Promise<ClassifyResult> {
  const shortDocChunks = options?.shortDocChunks ?? DEFAULT_SHORT_DOC_CHUNKS;
  const coalesceThreshold = options?.coalesceOtherThreshold ?? DEFAULT_COALESCE_OTHER_THRESHOLD;

  if (chunks.length === 0) {
    return { sections: [], corrections: 0, classifierSkipped: false };
  }

  if (shortDocChunks > 0 && chunks.length <= shortDocChunks) {
    return {
      sections: [fallbackSection(chunks)],
      corrections: 0,
      classifierSkipped: true,
    };
  }

  const validTypeIds = new Set(
    types
      .filter((t) => typeof t === "object" && t !== null && typeof t.id === "string" && t.id.trim())
      .map((t) => t.id!),
  );

  const prompt = buildClassifierPrompt(chunks, types);

  let raw: string | null = null;
  let error: string | null = null;

  try {
    raw = await provider.generate(prompt, true);
  } catch (exc: unknown) {
    const name = exc instanceof Error ? exc.constructor.name : "Error";
    const msg = exc instanceof Error ? exc.message : String(exc);
    error = `${name}: ${msg}`;
    console.log(`[koji-classify] Classifier error: ${error} — falling back to whole-document section`);
  }

  const parsed = raw !== null ? parseClassifierJson(raw) : null;
  let [sections, corrections] = normalizeClassifierResponse(parsed, chunks.length, validTypeIds, chunks);

  let coalesced: [Section[], string | null];
  coalesced = coalesceOtherSections(sections, chunks.length, coalesceThreshold, chunks);
  sections = coalesced[0];

  return {
    sections,
    corrections,
    classifierSkipped: false,
  };
}
